### ISSUE TITLE
no coverage for terminated processes

### ISSUE THREAD

**Issue by sezanzeb** [coveragepy/coveragepy#1069]
The following example won't give me coverage for the process

```py
import time
import unittest
import multiprocessing


def target():
    a = 1
    while a < 10:
        time.sleep(0.1)
        a += 1
    return a


def spawn():
    process = multiprocessing.Process(target=target)
    process.start()
    return process


class TestAsdf(unittest.TestCase):
    def tearDown(self):
        self.p.terminate()

    def test_asdf(self):
        self.p = spawn()


if __name__ == "__main__":
    unittest.main()
```

```
[run]
source = tests/testcases/
concurrency = multiprocessing
```

```
coverage run tests/testcases/asdf.py && coverage combine && coverage report -m
.
----------------------------------------------------------------------
Ran 1 test in 0.001s

OK
Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
tests/testcases/__init__.py          0      0   100%
tests/testcases/asdf.py             20      5    75%   7-11
```

However, when removing the tearDown function, it catches it:

```
Name                             Stmts   Miss  Cover   Missing
--------------------------------------------------------------
tests/testcases/__init__.py          0      0   100%
tests/testcases/asdf.py             18      0   100%
```

```
-- sys -------------------------------------------------------
                        version: 5.3
                       coverage: /usr/lib/python3.8/site-packages/coverage/__init__.py
                         tracer: -none-
                        CTracer: available
           plugins.file_tracers: -none-
            plugins.configurers: -none-
      plugins.context_switchers: -none-
              configs_attempted: /mnt/data/Code/setxkbmap-gtk/.coveragerc
                   configs_read: /mnt/data/Code/setxkbmap-gtk/.coveragerc
                    config_file: /mnt/data/Code/setxkbmap-gtk/.coveragerc
                config_contents: b'[run]\nsource = tests/testcases/\nconcurrency = multiprocessing\n'
                      data_file: -none-
                         python: 3.8.5 (default, Sep  5 2020, 10:50:12) [GCC 10.2.0]
                       platform: Linux-5.4.72-1-MANJARO-x86_64-with-glibc2.2.5
                 implementation: CPython
                     executable: /usr/bin/python
                   def_encoding: utf-8
                    fs_encoding: utf-8
                            pid: 7240
                            cwd: /mnt/data/Code/setxkbmap-gtk
                           path: /usr/bin
                                 /usr/lib/python38.zip
                                 /usr/lib/python3.8
                                 /usr/lib/python3.8/lib-dynload
                                 /home/mango/.local/lib/python3.8/site-packages
                                 /usr/lib/python3.8/site-packages
                                 /usr/lib/python3.8/site-packages/evdev-1.3.0-py3.8-linux-x86_64.egg
                                 /usr/lib/python3.8/site-packages/coverage_badge-1.0.1-py3.8.egg
                                 /usr/lib/python3.8/site-packages/dbus_python-1.2.16-py3.8-linux-x86_64.egg
                                 /mnt/data/Code/setxkbmap-gtk
                    environment: COVERAGE_PROCESS_START = .coveragerc
                                 COVERAGE_RCFILE = /mnt/data/Code/setxkbmap-gtk/.coveragerc
                   command_line: /usr/bin/coverage debug sys
                sqlite3_version: 2.6.0
         sqlite3_sqlite_version: 3.33.0
             sqlite3_temp_store: 0
        sqlite3_compile_options: COMPILER=gcc-10.2.0
                                 ENABLE_COLUMN_METADATA
                                 ENABLE_DBSTAT_VTAB
                                 ENABLE_FTS3
                                 ENABLE_FTS3_TOKENIZER
                                 ENABLE_FTS4
                                 ENABLE_FTS5
                                 ENABLE_JSON1
                                 ENABLE_RTREE
                                 ENABLE_UNLOCK_NOTIFY
                                 HAVE_ISNAN
                                 MAX_EXPR_DEPTH=10000
                                 MAX_VARIABLE_NUMBER=250000
                                 SECURE_DELETE
                                 TEMP_STORE=1
                                 THREADSAFE=1
```

**Comment by sezanzeb** [comment]
is it because this never happens: https://github.com/nedbat/coveragepy/blob/7f8dd68dd6da360374603f3a352c07713a3b082d/coverage/multiproc.py#L51 ?

I guess I need to figure out a way to stop my endless loop without just terminating the process. With a message over a pipe or something probably.

**Comment by sezanzeb** [comment]
the problem with this seems indeed, that the process is killed early. putting `debug = multiproc` into the .coveragerc helped. The `Saved multiprocessing data` message was missing.

After searching for some more time I found another .terminate in my code in a destructor that I forgot about. It seems to work now. I have added a pipe that sends a message to my process to stop iterating. I just don't use .terminate anymore.
